UT RuntimeError: MKL FFT doesn't support tensor of type

### 🐛 Describe the bug

8 UTs fail on Windows  and Linux with Intel XPU (Arc B580) with `RuntimeError: MKL FFT doesn't support tensor of type`.


### Affected Test Cases

Cases:
op_ut,third_party.torch-xpu-ops.test.xpu.test_meta_xpu.TestMetaXPU,test_dispatch_meta_outplace_stft_xpu_bfloat16
op_ut,third_party.torch-xpu-ops.test.xpu.test_meta_xpu.TestMetaXPU,test_dispatch_symbolic_meta_outplace_stft_xpu_bfloat16
op_ut,third_party.torch-xpu-ops.test.xpu.test_meta_xpu.TestMetaXPU,test_meta_outplace_stft_xpu_bfloat16
op_ut,third_party.torch-xpu-ops.test.xpu.test_meta_xpu.TestMetaXPU,test_meta_outplace_stft_xpu_bfloat16
op_ut,third_party.torch-xpu-ops.test.xpu.test_ops_xpu.TestCommonXPU,test_dtypes_stft_xpu
op_ut,third_party.torch-xpu-ops.test.xpu.test_ops_xpu.TestCommonXPU,test_python_ref_executor__refs_stft_executor_aten_xpu_bfloat16
op_ut,third_party.torch-xpu-ops.test.xpu.test_ops_xpu.TestCommonXPU,test_python_ref_torch_fallback__refs_stft_xpu_bfloat16
op_ut,third_party.torch-xpu-ops.test.xpu.test_schema_check.TestSchemaCheckModeOpInfoXPU,test_schema_correctness_stft_xpu_bfloat16

### Error Message

<details>
<summary>test_schema_correctness_stft_xpu_bfloat16 </summary>

```
___ TestSchemaCheckModeOpInfoXPU.test_schema_correctness_stft_xpu_bfloat16 ____
[gw1] win32 -- Python 3.12.13 C:\Users\gta\miniforge3\envs\202607060305_viable-s_32.0.101.8860_2026.0.0.644\python.exe
Traceback (most recent call last):
  File "C:\Users\gta\miniforge3\envs\202607060305_viable-s_32.0.101.8860_2026.0.0.644\Lib\site-packages\torch\testing\_internal\common_device_type.py", line 1450, in test_wrapper
    return test(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\gta\miniforge3\envs\202607060305_viable-s_32.0.101.8860_2026.0.0.644\Lib\site-packages\torch\testing\_internal\common_utils.py", line 2647, in wrapper
    fn(*args, **kwargs)
  File "C:\Users\gta\repositories\pytorch\pytorch\third_party\torch-xpu-ops\test\xpu\test_schema_check.py", line 543, in test_schema_correctness
    op(sample.input, *sample.args, **sample.kwargs)
  File "C:\Users\gta\miniforge3\envs\202607060305_viable-s_32.0.101.8860_2026.0.0.644\Lib\site-packages\torch\testing\_internal\opinfo\core.py", line 1241, in __call__
    return self.op(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\gta\miniforge3\envs\202607060305_viable-s_32.0.101.8860_2026.0.0.644\Lib\site-packages\torch\functional.py", line 680, in stft
    return _VF.stft(  # type: ignore[attr-defined]
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\gta\miniforge3\envs\202607060305_viable-s_32.0.101.8860_2026.0.0.644\Lib\site-packages\torch\_compile.py", line 54, in inner
    return disable_fn(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\gta\miniforge3\envs\202607060305_viable-s_32.0.101.8860_2026.0.0.644\Lib\site-packages\torch\_dynamo\eval_frame.py", line 1468, in _fn
    return fn(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^
  File "C:\Users\gta\miniforge3\envs\202607060305_viable-s_32.0.101.8860_2026.0.0.644\Lib\site-packages\torch\_subclasses\schema_check_mode.py", line 184, in __torch_dispatch__
    out = func(*args, **kwargs)
          ^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\gta\miniforge3\envs\202607060305_viable-s_32.0.101.8860_2026.0.0.644\Lib\site-packages\torch\_ops.py", line 876, in __call__
    return self._op(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^
RuntimeError: MKL FFT doesn't support tensor of type

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "C:\Users\gta\miniforge3\envs\202607060305_viable-s_32.0.101.8860_2026.0.0.644\Lib\unittest\case.py", line 58, in testPartExecutor
    yield
  File "C:\Users\gta\miniforge3\envs\202607060305_viable-s_32.0.101.8860_2026.0.0.644\Lib\unittest\case.py", line 634, in run
    self._callTestMethod(testMethod)
  File "C:\Users\gta\miniforge3\envs\202607060305_viable-s_32.0.101.8860_2026.0.0.644\Lib\unittest\case.py", line 589, in _callTestMethod
    if method() is not None:
       ^^^^^^^^
  File "C:\Users\gta\miniforge3\envs\202607060305_viable-s_32.0.101.8860_2026.0.0.644\Lib\site-packages\torch\testing\_internal\common_utils.py", line 3590, in wrapper
    method(*args, **kwargs)
  File "C:\Users\gta\miniforge3\envs\202607060305_viable-s_32.0.101.8860_2026.0.0.644\Lib\site-packages\torch\testing\_internal\common_device_type.py", line 627, in instantiated_test
    result = test(self, **param_kwargs)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\gta\miniforge3\envs\202607060305_viable-s_32.0.101.8860_2026.0.0.644\Lib\site-packages\torch\testing\_internal\common_device_type.py", line 1534, in dep_fn
    return fn(slf, *args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\gta\miniforge3\envs\202607060305_viable-s_32.0.101.8860_2026.0.0.644\Lib\site-packages\torch\testing\_internal\common_utils.py", line 1933, in wrapper
    fn(*args, **kwargs)
  File "C:\Users\gta\miniforge3\envs\202607060305_viable-s_32.0.101.8860_2026.0.0.644\Lib\site-packages\torch\testing\_internal\common_device_type.py", line 1462, in test_wrapper
    raise e_tracked from e
Exception: MKL FFT doesn't support tensor of type

Caused by sample input at index 0: SampleInput(input=Tensor[size=(100,), device="xpu:0", dtype=torch.bfloat16], args=(), kwargs={'n_fft': '10', 'return_complex': 'True'}, broadcasts_input=False, name='')

```

</details>


<details>
<summary>test_meta_outplace_stft_xpu_bfloat16</summary>

```
______________ TestMetaXPU.test_meta_outplace_stft_xpu_bfloat16 _______________
[gw4] win32 -- Python 3.12.13 C:\Users\gta\miniforge3\envs\202607060305_viable-s_32.0.101.8860_2026.0.0.644\python.exe
Traceback (most recent call last):
  File "C:\Users\gta\repositories\pytorch\pytorch\third_party\torch-xpu-ops\test\xpu\test_meta_xpu.py", line 629, in run_meta_crossref
    rs = func(*args, **kwargs)
         ^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\gta\miniforge3\envs\202607060305_viable-s_32.0.101.8860_2026.0.0.644\Lib\site-packages\torch\functional.py", line 680, in stft
    return _VF.stft(  # type: ignore[attr-defined]
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
RuntimeError: MKL FFT doesn't support tensor of type

```

</details>

<details>
<summary>test_dtypes_stft_xpu </summary>

```
______________________ TestCommonXPU.test_dtypes_stft_xpu ______________________
[gw9] linux -- Python 3.10.20 /tmp/xpu-tool/Python/3.10.20/x64/bin/python
Traceback (most recent call last):
  File "/tmp/xpu-tool/Python/3.10.20/x64/lib/python3.10/site-packages/torch/testing/_internal/common_device_type.py", line 1450, in test_wrapper
    return test(*args, **kwargs)
  File "/tmp/xpu-tool/Python/3.10.20/x64/lib/python3.10/site-packages/torch/testing/_internal/common_device_type.py", line 1534, in dep_fn
    return fn(slf, *args, **kwargs)
  File "/tmp/xpu-tool/Python/3.10.20/x64/lib/python3.10/site-packages/torch/testing/_internal/common_device_type.py", line 1812, in only_fn
    return fn(self, *args, **kwargs)
  File "/__w/torch-xpu-ops/torch-xpu-ops/pytorch/third_party/torch-xpu-ops/test/xpu/../../../../test/test_ops.py", line 1810, in test_dtypes
    self.fail(msg)
  File "/tmp/xpu-tool/Python/3.10.20/x64/lib/python3.10/unittest/case.py", line 675, in fail
    raise self.failureException(msg)
AssertionError: The supported dtypes for stft on device type xpu are incorrect!
The following dtypes did not work in forward but are listed by the OpInfo: {torch.bfloat16}.
The following dtypes did not work in backward but are listed by the OpInfo: {torch.bfloat16}.
Unexpected failures raised the following errors:
torch.bfloat16 - MKL FFT doesn't support tensor of type


The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/tmp/xpu-tool/Python/3.10.20/x64/lib/python3.10/site-packages/torch/testing/_internal/common_utils.py", line 3590, in wrapper
    method(*args, **kwargs)
  File "/tmp/xpu-tool/Python/3.10.20/x64/lib/python3.10/site-packages/torch/testing/_internal/common_device_type.py", line 627, in instantiated_test
    result = test(self, **param_kwargs)
  File "/tmp/xpu-tool/Python/3.10.20/x64/lib/python3.10/site-packages/torch/testing/_internal/common_device_type.py", line 1534, in dep_fn
    return fn(slf, *args, **kwargs)
  File "/tmp/xpu-tool/Python/3.10.20/x64/lib/python3.10/site-packages/torch/testing/_internal/common_utils.py", line 1933, in wrapper
    fn(*args, **kwargs)
  File "/tmp/xpu-tool/Python/3.10.20/x64/lib/python3.10/site-packages/torch/testing/_internal/common_device_type.py", line 1462, in test_wrapper
    raise e_tracked from e
Exception: The supported dtypes for stft on device type xpu are incorrect!
The following dtypes did not work in forward but are listed by the OpInfo: {torch.bfloat16}.
The following dtypes did not work in backward but are listed by the OpInfo: {torch.bfloat16}.
Unexpected failures raised the following errors:
torch.bfloat16 - MKL FFT doesn't support tensor of type

```

</details>


### Versions

<details>
<summary>Click to expand environment details</summary>

```
PyTorch version: 2.14.0a0+git3043bfb
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: Microsoft Windows 11 Pro (10.0.26100 64-bit)
GCC version: Could not collect
Clang version: Could not collect
CMake version: version 3.31.6
Libc version: N/A

Python version: 3.12.13 | packaged by conda-forge | (main, Mar  5 2026, 16:36:12) [MSC v.1944 64 bit (AMD64)] (64-bit runtime)
Python platform: Windows-11-10.0.26100-SP0
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
Is XPU available: True
XPU used to build PyTorch: 20260000
Intel GPU driver version:
* 32.0.101.8860 (20260625000000.******+***)
Intel GPU models onboard:
* Intel(R) Arc(TM) B580 Graphics
Intel GPU models detected:
* [0] _XpuDeviceProperties(name='Intel(R) Arc(TM) B580 Graphics', platform_name='Intel(R) oneAPI Unified Runtime over Level-Zero V2', type='gpu', device_id=0xE20B, uuid=86800be2-0000-0000-0300-000000000000, driver_version='1.15.38308', total_memory=11875MB, local_mem_size=128KB, last_level_cache_size=18432KB, max_compute_units=160, memory_clock_rate=0MHz, memory_bus_width=64-bit, gpu_eu_count=160, gpu_subslice_count=20, max_work_group_size=1024, max_num_sub_groups=64, sub_group_sizes=[16 32], has_fp16=1, has_fp64=1, has_atomic64=1, is_integrated_gpu=0)
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: False
Caching allocator config: N/A

CPU:
Name: 13th Gen Intel(R) Core(TM) i9-13900
Manufacturer: GenuineIntel
Family: 207
Architecture: 9
ProcessorType: 3
DeviceID: CPU0
CurrentClockSpeed: 2000
MaxClockSpeed: 2000
L2CacheSize: 32768
L2CacheSpeed: None
Revision: None

Versions of relevant libraries:
[pip3] bert_pytorch==0.0.1a4
[pip3] functorch==1.14.0a0+b71aa0b
[pip3] intel-openmp==2026.1.0
[pip3] mkl-include==2026.0.0
[pip3] mkl-static==2026.0.0
[pip3] mypy==2.1.0
[pip3] mypy_extensions==1.1.0
[pip3] numpy==1.26.4
[pip3] onemkl-license==2026.0.0
[pip3] onnx==1.21.0
[pip3] onnx-ir==0.1.16
[pip3] onnxscript==0.6.2
[pip3] optree==0.13.0
[pip3] pytorch-labs-segment-anything-fast==0.2
[pip3] tbb==2023.0.0
[pip3] tbb-devel==2023.0.0
[pip3] tcmlib==1.5.0
[pip3] torch==2.14.0a0+git3043bfb
[pip3] torch_geometric==2.4.0
[pip3] torchao==0.17.0
[pip3] torchaudio==2.11.0a0+c0cbdb9
[pip3] torchbench==0.1
[pip3] torchmetrics==1.9.0
[pip3] torchmultimodal==0.1.0b0
[pip3] torchrec-nightly==2022.4.26
[pip3] torchvision==0.29.0a0+0bc41e6
[pip3] torchx-nightly==2026.6.29
[pip3] triton-xpu==3.7.2+git5fcc14d9
[conda] bert-pytorch              0.0.1a4                   dev_0    <develop>
[conda] functorch                 1.14.0a0+b71aa0b          pypi_0    pypi
[conda] intel-openmp              2026.1.0                 pypi_0    pypi
[conda] mkl-include               2026.0.0                 pypi_0    pypi
[conda] mkl-static                2026.0.0                 pypi_0    pypi
[conda] numpy                     1.26.4                   pypi_0    pypi
[conda] onemkl-license            2026.0.0                 pypi_0    pypi
[conda] optree                    0.13.0                   pypi_0    pypi
[conda] pytorch-labs-segment-anything-fast 0.2                      pypi_0    pypi
[conda] tbb                       2023.0.0                 pypi_0    pypi
[conda] tbb-devel                 2023.0.0                 pypi_0    pypi
[conda] tcmlib                    1.5.0                    pypi_0    pypi
[conda] torch                     2.14.0a0+git3043bfb          pypi_0    pypi
[conda] torch-geometric           2.4.0                    pypi_0    pypi
[conda] torchao                   0.17.0                   pypi_0    pypi
[conda] torchaudio                2.11.0a0+c0cbdb9          pypi_0    pypi
[conda] torchbench                0.1                      pypi_0    pypi
[conda] torchmetrics              1.9.0                    pypi_0    pypi
[conda] torchmultimodal           0.1.0b0                  pypi_0    pypi
[conda] torchrec-nightly          2022.4.26                pypi_0    pypi
[conda] torchvision               0.29.0a0+0bc41e6          pypi_0    pypi
[conda] torchx-nightly            2026.6.29                pypi_0    pypi
[conda] triton-xpu                3.7.2+git5fcc14d9          pypi_0    pypi
```

</details>

--- a/src/ATen/native/xpu/mkl/SpectralOps.cpp
+++ b/src/ATen/native/xpu/mkl/SpectralOps.cpp
@@ -321,7 +321,8 @@ const Tensor& _fft_apply_normalization(
 
 // TODO: Remove this work-around in future.
 Tensor promote_fft_input(const Tensor& input) {
-  if (input.scalar_type() == ScalarType::Half)
+  if (input.scalar_type() == ScalarType::Half ||
+      input.scalar_type() == ScalarType::BFloat16)
     return input.to(ScalarType::Float);
   if (input.scalar_type() == ScalarType::ComplexHalf)
     return input.to(ScalarType::ComplexFloat);

--- a/test/regressions/test_fft_r2c_bf16.py
+++ b/test/regressions/test_fft_r2c_bf16.py
@@ -1,0 +1,40 @@
+# Copyright 2026 Intel Corporation
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+
+# Owner(s): ["module: intel"]
+# Regression test for https://github.com/intel/torch-xpu-ops/issues/4279
+#
+# bfloat16 real-to-complex (R2C) FFT previously failed on XPU with
+# "MKL FFT doesn't support tensor of type". The XPU MKL backend now promotes
+# BFloat16 input to Float32 in promote_fft_input
+# (src/ATen/native/xpu/mkl/SpectralOps.cpp).
+#
+# This test monitors that the R2C output dtype does not silently change:
+# for a bfloat16 input the R2C FFT output must be ComplexFloat (complex64),
+# matching the CUDA contract (bf16 promotes to fp32, not fp16).
+
+import torch
+from torch.testing._internal.common_device_type import instantiate_device_type_tests
+from torch.testing._internal.common_utils import run_tests, TestCase
+
+
+class TestFftR2cBFloat16(TestCase):
+    def test_fft_r2c_bfloat16_output_dtype(self, device):
+        # torch.fft.rfft is the R2C FFT
+        x = torch.randn(100, dtype=torch.bfloat16, device=device)
+        out = torch.fft.rfft(x, n=10)
+        # bf16 promotes to fp32 -> R2C output must be complex64
+        self.assertEqual(out.dtype, torch.complex64)
+
+
+instantiate_device_type_tests(
+    TestFftR2cBFloat16, globals(), only_for="xpu", allow_xpu=True
+)
+
+if __name__ == "__main__":
+    run_tests()

